File: backend/app/schemas/lyrics.py
"""Schémas Pydantic pour Lyrics.

Deux formes de sortie distinctes, jamais mélangées :
- LyricsVisibilityRead : ce qu'un visiteur public / autre USER reçoit
  (GET /lyrics/song/{id}) — ne contient structurellement AUCUN champ
  interne (statut, source, détenteur des droits, auteur...).
- LyricsOwnerRead : vue enrichie réservée à l'auteur de la soumission
  et à l'ADMIN (POST /lyrics, PUT /lyrics/{id}, GET /lyrics/mine, et
  GET /lyrics/song/{id} lorsque l'appelant est l'auteur/ADMIN).

Cette séparation est structurelle (deux classes différentes), pas une
simple omission conditionnelle de champs à l'exécution — conforme au
principe déjà appliqué à LyricsPublicRead/LyricsAdminRead (Livrable 3
§18).
"""

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.song import LanguageBrief

SourceType = Literal[
    "ORIGINAL", "ARTIST", "RIGHTS_HOLDER", "LICENSE", "PARTNER", "PUBLIC_DOMAIN", "USER_SUBMITTED"
]


class LyricsVisibilityRead(BaseModel):
    """Vue publique — GET /lyrics/song/{song_id} pour un visiteur non
    autorisé à voir le contenu réel (non-auteur, non-ADMIN, ou statut
    != AUTHORIZED / expiré)."""

    available: bool
    language: LanguageBrief | None = None
    content: str | None = None


class LyricsOwnerRead(BaseModel):
    """Vue enrichie — auteur de la soumission ou ADMIN uniquement."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    song_id: UUID
    language: LanguageBrief
    content: str
    source_type: str
    source_url: str | None = None
    rights_holder: str | None = None
    authorization_status: str
    authorization_reference: str | None = None
    authorization_date: date | None = None
    expiration_date: date | None = None
    submitted_by_user_id: UUID | None = None
    reviewed_by_user_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class LyricsCreate(BaseModel):
    """Contrat de soumission — `submitted_by_user_id` et
    `authorization_status` sont structurellement absents : le client
    ne peut physiquement pas les fournir."""

    song_id: UUID
    language_id: UUID
    content: str = Field(min_length=1)
    source_type: SourceType
    source_url: str | None = None
    rights_holder: str | None = Field(default=None, max_length=255)


class LyricsUpdate(BaseModel):
    """Édition — auteur (si PENDING) ou ADMIN. Seuls `content`,
    `source_url`, `rights_holder` sont modifiables ; `song_id`,
    `language_id`, `source_type`, `submitted_by_user_id`,
    `authorization_status`, `reviewed_by_user_id` sont structurellement
    absents de ce schéma."""

    content: str | None = Field(default=None, min_length=1)
    source_url: str | None = None
    rights_holder: str | None = Field(default=None, max_length=255)

File: frontend/types/lyrics.ts
import type { LanguageBrief } from "@/types/catalog";

/** lyrics.source_type CHECK — backend/app/schemas/lyrics.py (SourceType). */
export type LyricsSourceType =
  | "ORIGINAL"
  | "ARTIST"
  | "RIGHTS_HOLDER"
  | "LICENSE"
  | "PARTNER"
  | "PUBLIC_DOMAIN"
  | "USER_SUBMITTED";

/** authorization_status CHECK, partagé lyrics/translations/rights_records
 * — docs/04-database/database.md. EXPIRED est un statut effectif calculé,
 * jamais littéralement écrit en base, mais reste une valeur possible sur
 * le fil (ex. GET /lyrics/mine d'un auteur). */
export type AuthorizationStatus = "PENDING" | "AUTHORIZED" | "REJECTED" | "EXPIRED" | "REVOKED";

/**
 * Vue publique — GET /lyrics/song/{song_id} pour un visiteur non
 * autorisé à voir le contenu réel (non-auteur, non-ADMIN, ou statut
 * != AUTHORIZED / expiré). Structurellement séparée de LyricsOwnerRead :
 * ne contient AUCUN champ interne (statut, source, droits...).
 */
export interface LyricsVisibilityRead {
  available: boolean;
  language: LanguageBrief | null;
  content: string | null;
}

/** Vue enrichie — auteur de la soumission ou ADMIN uniquement. */
export interface LyricsOwnerRead {
  id: string;
  song_id: string;
  language: LanguageBrief;
  content: string;
  source_type: LyricsSourceType;
  source_url: string | null;
  rights_holder: string | null;
  authorization_status: AuthorizationStatus;
  authorization_reference: string | null;
  authorization_date: string | null;
  expiration_date: string | null;
  submitted_by_user_id: string | null;
  reviewed_by_user_id: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * GET /lyrics/song/{song_id} renvoie l'une ou l'autre forme selon le
 * viewer (backend/app/services/lyrics_service.py:get_visibility), sans
 * discriminant explicite sur le fil JSON. Utiliser isLyricsOwnerView
 * pour distinguer les deux à l'exécution.
 */
export type LyricsSongView = LyricsVisibilityRead | LyricsOwnerRead;

export function isLyricsOwnerView(view: LyricsSongView): view is LyricsOwnerRead {
  return "authorization_status" in view;
}

/** `submitted_by_user_id`/`authorization_status` structurellement absents
 * — forcés côté serveur, jamais fournis par le client. */
export interface LyricsCreate {
  song_id: string;
  language_id: string;
  content: string;
  source_type: LyricsSourceType;
  source_url?: string | null;
  rights_holder?: string | null;
}

/** Édition — uniquement le contenu, `song_id`/`language_id`/`source_type`/
 * `submitted_by_user_id`/`authorization_status`/`reviewed_by_user_id`
 * structurellement absents. */
export interface LyricsUpdate {
  content?: string;
  source_url?: string | null;
  rights_holder?: string | null;
}
